### ISSUE TITLE
"Return Null When Not Found" boolean for "Get Object Root" node

### ISSUE THREAD

**Issue by amplified1** [Yellow-Dog-Man/Resonite-Issues#1483]
### Is your feature request related to a problem? Please describe.

The "Get Object Root" node cannot actually verify the existence of an object root as it simply returns the input slot when there is no valid object root, which is indistinguishable from if the slot simply happened to have an object root on the same slot that has been analyzed.

### Describe the solution you'd like

Add a boolean input to the node that makes the node simply return a null slot when there is no valid object root to be found in the hierarchy.

### Describe alternatives you've considered

Another way this could be achieved is having a "isFound" boolean on the output of the node.

Technically you can verify object roots in the current game, but it requires a clunky workaround of traversing to a child of a slot, or adding one if there isn't already, and checking the object root once more from there, not ideal.

### Requesters

Amplified, alizard

**Comment by JackTheFoxOtter** [comment]
I think the boolean output for wether or not a (implicit or explicit) object root was found is the better approach. I agree that it would be useful to be able to differenciate this. That said, does it also return the input slot reference if you pass true to "only explicit"?

**Comment by amplified1** [comment]
> I think the boolean output for wether or not a (implicit or explicit) object root was found is the better approach. I agree that it would be useful to be able to differenciate this. That said, does it also return the input slot reference if you pass true to "only explicit"?

Yes it still returns it when set to only explicit unfortunately, and also I do agree, as I wrote the issue I was slowly leaning more towards the boolean output being a better option, and I think that does solidify it.

**Comment by Frooxius** [comment]
I'm not sure about making this change, because this is kinda in conflict with what the engine considers an "ObjectRoot" and it's based on a false premise.

In absence of any ObjectRoot or IObjectRoot components in the parent hierarchy, the object itself is considered its own implicit ObjectRoot, therefore it's returned, so the node's behavior follows that logic - there's no concept of "Not Found", because the object root is found - it's the object itself.

If we diverge it from that, we create an odd situation where you have a node that doesn't match the rest of the engine's behavior, which creates an inconsistency.

My question is, what are your use-cases where you need to know if the ObjectRoot or IObjectRoot component is specifically present?

**Comment by JackTheFoxOtter** [comment]
Îf the slot itself is considers it's own *implicit* object root under absence of other options, it shouldn't be returned when you pass true to "Only Explicit" though.

**Comment by Frooxius** [comment]
That's not what the Only Explicit means in this context.

This differentiates between specifically ObjectRoot (which is explicit marking) vs implicitly marked object root (e.g. grabbables). In absence of both though, the object is still its own object root.

**Comment by gentlecolts** [comment]
I think there's a disconnect here between the explicit "object root" component, and the concept of an object's root.  For the former, it's very possible that, given a starting slot, either that slot has an object root component, or there's no such component on any of its parents up to the root slot.  For the latter, however, there can't really ever be a situation where there's "no object root", because either the slot itself is the conceptual root of the object, or, in the absence of explicit labels, you eventually reach the scene root, which is still _an_ object root.

I think in this case the requester is asking for something explicitly looking for the object root component, or the lack of one, regardless of other situations which might potentially be conceptually (and implicitly) the root of an object, which I think may be a valuable to make some distinction for.

I also think that distinction could be made explicitly by users once we have component access and collections (ie, we could use a hypothetical "get components by type" node to check up the heirarchy), but it may still be useful to have this behavior encapsulated in a node if it's widespread enough

**Comment by JackTheFoxOtter** [comment]
I believe the way to achieve this through generic component access in the future is probably the way to go. I got confused, then thought about it for a bit, and have come to the same conclusion. A specific node to find the presence of a single specific component is likely not gonna be very useful, and in the case of object root, might cause people to "misuse" the object root tagging functionality.

**Comment by ProbablePrime** [comment]
Could always output the type as some form of Enum? Implicit, Explicit, etc

**Comment by Frooxius** [comment]
I agree that there is a disconnect - but that's where the issue lies. You are asking us to consider a different interpretation of ObjectRoots with this, than the rest of the engine uses.

My concern is that if we do this, it will end up being more confusing in the long run, as this node will use one interpretation, while other parts of engine will steal treat everything with the other definition and behave differently.

Generally I try to keep the nodes behavior to match the behavior of the internal API's to keep things consistent.

And yes, generic component access would let you just find the component itself, rather than what the engine considers "object root". 

It would also help to get answer to my question above: What is your use-case for this? Helping us understand why you'd need this change would help us understand the problem better. Without that, it'll just be hitting head against the wall.

**Comment by amplified1** [comment]
> It would also help to get answer to my question above: What is your use-case for this? Helping us understand why you'd need this change would help us understand the problem better. Without that, it'll just be hitting head against the wall.

The use case I had in mind is to limit systems that are made to interact with arbitrary objects to only interact with the slot that has been defined as the root of said object by the creator, say you want to make some sort of gravity gun object, reparenting the first thing you raycast would be extremely dangerous in some cases, grabbing random colliders devoid of visuals and such. With currently available methods, an object like this would not be able to differentiate something that has been defined as an object and something that is just the first slot it can find. I feel like it would make sense to have some way to differentiate component object roots, because you can *technically* do it now, it just requires a roundabout method that involves traversing to a child slot, and using the node again to check if you travel upwards. The usefulness of being able to see this is just that you can tell dynamically if a slot was intended by its creator to be a distinct object.
